File: platform/polyglot/langgraph-flow-serializer/mjs/src/exporters/mermaid.exporter.ts
/**
 * Mermaid flowchart exporter.
 *
 * Generates Mermaid `flowchart TD` syntax from an AIWorkflow.
 * Category-based node shapes:
 *   agents       → hexagon     {{label}}
 *   routing      → diamond     {label}
 *   conditions   → diamond     {label}
 *   prompts      → trapezoid   [/label\]
 *   models       → stadium     ([label])
 *   memory       → subroutine  [[label]]
 *   tools        → asymmetric  >label]
 *   (default)    → rectangle   [label]
 *
 * Nodes are grouped into Mermaid subgraphs by category.
 * Edges are rendered as labeled arrows.
 */

import type { AIWorkflow, AINode, AIEdge } from "../types.js";

type ShapeRenderer = (id: string, label: string) => string;

const SHAPE_MAP: Record<string, ShapeRenderer> = {
  agent: (id, label) => `${id}{{${escapeMermaid(label)}}}`,
  agents: (id, label) => `${id}{{${escapeMermaid(label)}}}`,
  routing: (id, label) => `${id}{${escapeMermaid(label)}}`,
  conditions: (id, label) => `${id}{${escapeMermaid(label)}}`,
  condition: (id, label) => `${id}{${escapeMermaid(label)}}`,
  prompt: (id, label) => `${id}[/${escapeMermaid(label)}\\]`,
  prompts: (id, label) => `${id}[/${escapeMermaid(label)}\\]`,
  model: (id, label) => `${id}([${escapeMermaid(label)}])`,
  models: (id, label) => `${id}([${escapeMermaid(label)}])`,
  "chat models": (id, label) => `${id}([${escapeMermaid(label)}])`,
  memory: (id, label) => `${id}[[${escapeMermaid(label)}]]`,
  tool: (id, label) => `${id}>${escapeMermaid(label)}]`,
  tools: (id, label) => `${id}>${escapeMermaid(label)}]`,
  control: (id, label) => `${id}((${escapeMermaid(label)}))`,
};

function escapeMermaid(text: string): string {
  // Escape quotes and pipe characters that have special meaning in Mermaid
  return text.replace(/"/g, "&quot;").replace(/\|/g, "&#124;");
}

function safeMermaidId(id: string): string {
  // Mermaid node IDs must be alphanumeric + underscores
  return id.replace(/[^a-zA-Z0-9_]/g, "_");
}

function renderNode(node: AINode): string {
  const safeId = safeMermaidId(node.id);
  const label = node.data.name ?? node.data.nodeType ?? node.id;
  const categoryKey = node.data.category.toLowerCase();
  const renderer = SHAPE_MAP[categoryKey];
  return renderer ? renderer(safeId, label) : `${safeId}[${escapeMermaid(label)}]`;
}

function renderEdge(edge: AIEdge, idMap: Map<string, string>): string {
  const src = idMap.get(edge.source) ?? safeMermaidId(edge.source);
  const tgt = idMap.get(edge.target) ?? safeMermaidId(edge.target);
  // Include handle label if it carries meaning beyond the default pattern
  const handleLabel =
    edge.sourceHandle && !edge.sourceHandle.endsWith("-output")
      ? edge.sourceHandle
      : null;
  return handleLabel ? `${src} -->|${escapeMermaid(handleLabel)}| ${tgt}` : `${src} --> ${tgt}`;
}

/**
 * Convert an AIWorkflow to a Mermaid flowchart string.
 *
 * @param workflow - The workflow to convert.
 * @returns A Mermaid `flowchart TD` diagram as a string.
 */
export function toMermaid(workflow: AIWorkflow): string {
  const lines: string[] = ["flowchart TD"];

  // Build ID map (original → safe Mermaid ID) for edge rendering
  const idMap = new Map<string, string>();
  for (const node of workflow.nodes) {
    idMap.set(node.id, safeMermaidId(node.id));
  }

  // Group nodes by category for subgraph generation
  const byCategory = new Map<string, AINode[]>();
  for (const node of workflow.nodes) {
    const cat = node.data.category;
    if (!byCategory.has(cat)) byCategory.set(cat, []);
    byCategory.get(cat)!.push(node);
  }

  // Render subgraphs
  for (const [category, nodes] of byCategory.entries()) {
    const subgraphId = safeMermaidId(category);
    lines.push(`  subgraph ${subgraphId}["${escapeMermaid(category)}"]`);
    for (const node of nodes) {
      lines.push(`    ${renderNode(node)}`);
    }
    lines.push("  end");
  }

  // Render edges
  for (const edge of workflow.edges as AIEdge[]) {
    lines.push(`  ${renderEdge(edge, idMap)}`);
  }

  return lines.join("\n");
}

File: platform/polyglot/langgraph-flow-serializer/py/langgraph_flow_serializer/exporters/mermaid_exporter.py
"""
Mermaid flowchart exporter.

Mirrors mjs/src/exporters/mermaid.exporter.ts.

Generates Mermaid `flowchart TD` syntax from an AIWorkflow.

Category-based node shapes:
  agents / agent     -> hexagon     {{label}}
  routing / conditions / condition -> diamond {label}
  prompt / prompts   -> trapezoid   [/label\\]
  model / models / chat models -> stadium ([label])
  memory             -> subroutine  [[label]]
  tool / tools       -> asymmetric  >label]
  control            -> circle      ((label))
  (default)          -> rectangle   [label]

Nodes are grouped into Mermaid subgraphs by category.
Edges are rendered as labeled arrows.
"""

from __future__ import annotations

import re
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from langgraph_flow_serializer.types import AIEdge, AINode, AIWorkflow


# Maps lowercase category keys to shape-rendering callables
def _hex(sid: str, label: str) -> str:
    return f'{sid}{{{{{_esc(label)}}}}}'

def _diamond(sid: str, label: str) -> str:
    return f'{sid}{{{_esc(label)}}}'

def _trapezoid(sid: str, label: str) -> str:
    return f'{sid}[/{_esc(label)}\\]'

def _stadium(sid: str, label: str) -> str:
    return f'{sid}([{_esc(label)}])'

def _subroutine(sid: str, label: str) -> str:
    return f'{sid}[[{_esc(label)}]]'

def _asymmetric(sid: str, label: str) -> str:
    return f'{sid}>{_esc(label)}]'

def _circle(sid: str, label: str) -> str:
    return f'{sid}(({_esc(label)}))'

def _rect(sid: str, label: str) -> str:
    return f'{sid}[{_esc(label)}]'


SHAPE_MAP: dict = {
    "agent": _hex,
    "agents": _hex,
    "routing": _diamond,
    "conditions": _diamond,
    "condition": _diamond,
    "prompt": _trapezoid,
    "prompts": _trapezoid,
    "model": _stadium,
    "models": _stadium,
    "chat models": _stadium,
    "memory": _subroutine,
    "tool": _asymmetric,
    "tools": _asymmetric,
    "control": _circle,
}


def _esc(text: str) -> str:
    """Escape characters that have special meaning in Mermaid labels."""
    return text.replace('"', "&quot;").replace("|", "&#124;")


def _safe_id(node_id: str) -> str:
    """Convert an arbitrary node ID to a valid Mermaid identifier."""
    return re.sub(r"[^a-zA-Z0-9_]", "_", node_id)


def _render_node(node: "AINode") -> str:
    safe = _safe_id(node.id)
    label = node.data.name or node.data.node_type or node.id
    renderer = SHAPE_MAP.get(node.data.category.lower(), _rect)
    return renderer(safe, label)


def _render_edge(edge: "AIEdge", id_map: dict[str, str]) -> str:
    src = id_map.get(edge.source, _safe_id(edge.source))
    tgt = id_map.get(edge.target, _safe_id(edge.target))
    # Include source handle label only if it's not the default "-output" pattern
    handle_label = (
        edge.source_handle
        if edge.source_handle and not edge.source_handle.endswith("-output")
        else None
    )
    if handle_label:
        return f"{src} -->|{_esc(handle_label)}| {tgt}"
    return f"{src} --> {tgt}"


def to_mermaid(workflow: "AIWorkflow") -> str:
    """
    Convert an AIWorkflow to a Mermaid flowchart string.

    Args:
        workflow: The workflow to convert.

    Returns:
        A Mermaid ``flowchart TD`` diagram as a string.
    """
    lines = ["flowchart TD"]

    # Build safe-ID map for edge rendering
    id_map = {node.id: _safe_id(node.id) for node in workflow.nodes}

    # Group nodes by category
    by_category: dict[str, list] = {}
    for node in workflow.nodes:
        by_category.setdefault(node.data.category, []).append(node)

    # Render subgraphs
    for category, nodes in by_category.items():
        subgraph_id = _safe_id(category)
        lines.append(f'  subgraph {subgraph_id}["{_esc(category)}"]')
        for node in nodes:
            lines.append(f"    {_render_node(node)}")
        lines.append("  end")

    # Render edges
    for edge in workflow.edges:
        lines.append(f"  {_render_edge(edge, id_map)}")

    return "\n".join(lines)
